src/action.c

```diff
@@ -56,6 +60,52 @@ const char *action_names[] = {
 	NULL
 };
 
+static char *
+action_str_from_arg(struct action_arg *arg)
+{
+	assert(arg->type == LAB_ACTION_ARG_STR);
+	return ((struct action_arg_str *)arg)->value;
+}
+
+static int
+action_int_from_arg(struct action_arg *arg)
+{
+	assert(arg->type == LAB_ACTION_ARG_INT);
+	return ((struct action_arg_int *)arg)->value;
+}
+
+static double
+action_double_from_arg(struct action_arg *arg)
+{
+	assert(arg->type == LAB_ACTION_ARG_DOUBLE);
+	return ((struct action_arg_double *)arg)->value;
+}
+
+static bool
+action_bool_from_arg(struct action_arg *arg)
+{
+	assert(arg->type == LAB_ACTION_ARG_BOOL);
+	return ((struct action_arg_bool *)arg)->value;
+}
+
+static struct wlr_box *
+action_box_from_arg(struct action_arg *arg)
+{
+	assert(arg->type == LAB_ACTION_ARG_BOX);
+	return &((struct action_arg_box *)arg)->value;
+}
+
+static struct action_arg *
+action_get_first_arg(struct action *action)
```

Comment:
In terms of the general design, what do you think of the following???

```
struct action {
	char *name;
	struct wl_list arguments;

	struct wl_list link; /* keybind.actions, etc */
}

struct action_arg {
	enum action_arg_type type;
	char *key;	/* e.g. "width", "direction" */
	union {
		char *value;
		int value;
		etc...
	}
	struct wl_list link; /* action.arguments */
}
```

It's possibly a bit simpler and behave more like a dictionary (although implemented as a `wl_list`).

The argument key gives the option of a helper in action.c to do things like `width = arg(action_arg_list, "width")`
It avoids functions like `get_first_argument`

Just a thought :smile:

Reply:
I like this better than the inheritance thing. Need to think a bit about this.

Re `width = arg(action_arg_list, "width")`:
How would the function signature for `arg()` look like in this case? At some point it would need to cast the result to one of the union members?

Reply:
Re union vs inheritance + casting, gave the union a try and ended up with this:
```c
struct action_arg {
        const char *key;      /* May be NULL if there is just one arg */
        struct wl_list link;  /* struct action.args */

        /* Private */
        uint32_t type;        /* enum action_arg_type */
        union {
                char *val_str;
                int val_int;
                double val_double;
                bool val_bool;
                struct wlr_box val_box;
        };
};
```
It makes the internal `action_x_from_arg()` function a bit easier to read but IMHO also
- exposes internal members to the outside
- make the struct look ugly

Are there technical reasons to prefer one over the other?

I can only think of a positive for the inheritance + casting approach:
- it only needs as much additional storage as required be the relevant argument instead of the max size of any argument